File: remedian.rb
class Remedian
  def initialize(base, ntile = 50)
    @base = base
    @exponent = 1
    @ntile = ntile
    @storage = Array.new
    @positions = Array.new
    @storage << Array.new
    @positions << 0
  end

  def self.quickselect(a, k)
    arr = a.dup
    loop do
      pivot = arr.delete_at(rand(arr.length))
      left, right = arr.partition { |x| x < pivot }
      if k == left.length
        return pivot
      elsif k < left.length
        arr = left
      else
        k = k - left.length - 1
        arr = right
      end
    end
  end

  def next(element, level = 0)
    if level >= @exponent
      @exponent = @exponent + 1
      @positions << 0
      @storage << Array.new
    end
    begin
      @storage[level][@positions[level]] = element
      if @positions[level] < @base - 1
        @positions[level] = @positions[level] + 1
      else
        @positions[level] = 0
        if level == 0
          self.next(Remedian::quickselect(@storage[level], (@storage[level].length * @ntile)/100), level + 1)
        else
          self.next(Remedian::quickselect(@storage[level], @storage[level].length/2), level + 1)
        end
      end
    rescue StandardError => e
      p 'error at %d' % element
      abort(e.message)

    end
  end

  def result()
    weighted_total = 0
    total_weight = 0
    weight = 1
    @storage.each do |a|
      a.each do |v|
        total_weight = total_weight + weight
        weighted_total = weighted_total + v*weight
      end
      weight = weight * @base
    end
    weighted_total/total_weight
  end
end

ntile = 50
set_size = 2000
base = 101

r = Remedian.new(base, ntile)

1.upto set_size do |num|
  r.next(num)
end

p 'sequential result: %d' % r.result
top = set_size
bottom = 1

r = Remedian.new(base, ntile)

loop do
6.times do
  r.next(top)
  top = top - 1
  break if top < bottom
end
break if top < bottom
5.times do
  r.next(bottom)
  bottom = bottom + 1
  break if top < bottom
end
break if top < bottom
end

p 'worst case result: %d' % r.result

p '10 random results:'

10.times do
  r = Remedian.new(base, ntile)
  a = *(1..set_size)
  while a.length > 0 do
    r.next(a.delete_at(rand(a.length)))
  end
  p r.result
end

p 'distribution of 10,000 random results:'

result_counts = Hash.new 0

10000.times do
  r = Remedian.new(base, ntile)
  a = *(1..set_size)
  while a.length > 0 do
    r.next(a.delete_at(rand(a.length)))
  end
  result_counts[r.result] += 1
end

result_counts.keys.sort.each do |k|
  p '%d: %d' % [k, result_counts[k]]
end

File: remedian.py
import random
from  collections import defaultdict
class Remedian:
    def __init__(self, base, ntile=50):
        self.ntile = ntile
        self.base = base
        self.exponent = 1
        self.storage = []
        self.positions = []
        self.storage.append([])
        self.positions.append(0)


    @staticmethod
    def partition(vector, left, right, pivotIndex):
        pivotValue = vector[pivotIndex]
        vector[pivotIndex], vector[right] = vector[right], vector[pivotIndex]  # Move pivot to end
        storeIndex = left
        for i in range(left, right):
            if vector[i] < pivotValue:
                vector[storeIndex], vector[i] = vector[i], vector[storeIndex]
                storeIndex += 1
        vector[right], vector[storeIndex] = vector[storeIndex], vector[right]  # Move pivot to its final place
        return storeIndex

    @staticmethod
    def _select(vector, left, right, k):
        while True:
            pivotIndex = random.randint(left, right)     # select pivotIndex between left and right
            pivotNewIndex = Remedian.partition(vector, left, right, pivotIndex)
            pivotDist = pivotNewIndex - left
            if pivotDist == k:
                return vector[pivotNewIndex]
            elif k < pivotDist:
                right = pivotNewIndex - 1
            else:
                k -= pivotDist + 1
                left = pivotNewIndex + 1

    @staticmethod
    def quickselect(vector, k, left=None, right=None):
        if left is None:
            left = 0
        lv1 = len(vector) - 1
        if right is None:
            right = lv1
        return Remedian._select(vector, left, right, k)

    def next_element(self, element, level=0):
        if level >= self.exponent:
            self.exponent += 1
            self.positions.append(0)
            self.storage.append([])
        if len(self.storage[level]) <= self.positions[level]:
            self.storage[level].append(element)
        else:
            self.storage[level][self.positions[level]] = element
        if self.positions[level] < (self.base - 1):
            self.positions[level] += 1
        else:
            self.positions[level] = 0
            if level == 0:
                self.next_element(Remedian.quickselect(self.storage[level], (len(self.storage[level]) * self.ntile)/100), level + 1)
            else:
                self.next_element(Remedian.quickselect(self.storage[level], len(self.storage[level])/2), level + 1)

    def result(self):
        weighted_total = 0
        total_weight = 0
        weight = 1
        for a in self.storage:
            for v in a:
                total_weight += weight
                weighted_total += v*weight
            weight *= self.base
        return weighted_total/total_weight

if __name__ == '__main__':
    ntile = 50
    set_size = 20000
    base = 101
    print "ntile of %d for a set size of %d using base %d" % (ntile, set_size, base)
    r = Remedian(base, ntile)

    for num in range(1,20000):
        r.next_element(num)

    print "sequential result: %d" % r.result()
    top = set_size
    bottom = 1
    r = Remedian(base, ntile)

    while(True):
        for i in range(base/2 + 1):
            r.next_element(top)
            top -= 1
            if top < bottom:
                break
        if top < bottom:
            break
        for i in range(base/2):
            r.next_element(bottom)
            bottom += 1
            if top < bottom:
                break
        if top < bottom:
            break

    print "worst case result: %d" % r.result()

    print "10 random permutations:"

    for i in range(10):
        r = Remedian(base, ntile)
        a = range(set_size)
        while len(a) > 0:
            r.next_element(a.pop(random.randint(0, len(a) - 1)))
        print r.result()


    print "distribution of 1000 random permutations:"
    result_counts = defaultdict(int)

    for i in range(1000):
        r = Remedian(base, ntile)
        a = range(set_size)
        while len(a) > 0:
            r.next_element(a.pop(random.randint(0, len(a) - 1)))
        result_counts[r.result()] += 1

    for key, value in sorted(result_counts.iteritems()):
        print "%d: %d" % (key,value)
